Library/Homebrew/livecheck/strategy.rb

```diff
@@ -80,6 +80,53 @@ def from_url(url, livecheck_strategy: nil, regex_provided: nil)
           (strategy.const_defined?(:PRIORITY) ? -strategy::PRIORITY : -DEFAULT_PRIORITY)
         end
       end
+
+      # Fetches the content at the URL and returns a hash containing the
+      # content and, if there are any redirections, the final URL.
+      # If `curl` encounters an error, the hash will contain a `:messages`
+      # array with the error message instead.
+      # @param url [String] the URL of the content to check
+      # @param user_agent [Symbol] the symbol for a corresponding user agent
+      # @return [Hash]
+      def self.page_content(url, user_agent: :default)
+        original_url = url
+
+        args = curl_args(
+          "--compressed", # Request/handle compressed responses
+          "--include",    # Include HTTP response headers in the output
+          "--location",   # Follow redirects
+          user_agent: user_agent,
+        )
+        result = curl_with_workarounds(*args, url, print_stdout: false, print_stderr: false, debug: false)
+
+        if result.status
+          /^(?<error_msg>curl: \(\d+\) .+)/ =~ result.stderr
+          return { messages: [error_msg] } if error_msg.present?
+        end
+
+        # stdout contains the header information followed by the page content.
+        # We use #scrub here to avoid "invalid byte sequence in UTF-8" errors.
+        output = result.stdout.scrub
+
+        # Separate the content from the headers
+        status_code = :unknown
+        while status_code == :unknown || status_code.to_s.start_with?("3")
+          headers, _, output = output.partition("\r\n\r\n")
+          status_code = headers[%r{HTTP/.* (\d+)}, 1]
+
+          location = headers[/^Location:\s*(.*)$/i, 1]
+          next unless location
+
+          location.chomp!
+          # Convert a relative redirect URL to an absolute URL
+          location = URI.join(url, location) unless location.match?(PageMatch::URL_MATCH_REGEX)
+          final_url = location
+        end
```

Comment:
I think this same logic, which also handles multiple redirects, also exists in `download_strategies.rb`.

Reply:
Perhaps extracting both to `utils/curl` would be good?

Reply:
There are similarities in what's happening in these three locations, so I tried my hand at creating a method that can handle these use cases (`#response_status_code_and_location`). I welcome feedback on the shape it takes, where it should exist, etc. but the current code is functional.

---

Basically, the work that I saw repeated in these areas was:

* Separating response headers from the body.
* Iterating through the headers for each of the responses to extract information.

The main functional difference between the approaches is whether the final `location` URL should be made absolute:

* [`Utils::Curl#curl_http_content_headers_and_checksum`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/curl.rb#L244-L250) seems to be geared toward returning headers in their original state, so it doesn't modify the `location` header URL.
* [`CurlDownloadStrategy#resolve_url_basename_time_file_size`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/download_strategy.rb#L425-L441) modifies the `location` header URL to make it absolute (using the original URL). This is more advanced than the simple `URI#join` statement I had in `Strategy#page_content`.

With this in mind, I made a method for separating the headers and body in cURL response text (`#response_headers_and_body`). This should work properly when the response contains both headers and body, only headers, and only the body.

If you need to work with the separated `headers`/`body` yourself, then you can execute this method before calling `#response_status_code_and_location` to extract the status_code and/or final location URL. If you don't need to separate the `headers`/`body`, then you can simply pass `#response_status_code_and_location` the response text and it will call `#response_headers_and_body` internally to format the separate the headers into an array.

---

Once this had gone through review, I'll add tests for these methods.